### PR TITLE
Fix concurrency issues for bools in ScriptEngine

### PR DESCRIPTION
**Pull request by huffman**
Several member variables were unsafely being accessed from multiple threads. Making these variables std::atomic makes this safe.

### CODE REVIEW

**Comment by huffman** [comment]
@imgntn When this builds could you see if reload-all is still crashing? It seems to be working for me now.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4dbfd3bfd936702e58ac90cc102d7df8bd4fe5a8 available at: https://gist.github.com/8e0d44e10fddb7f4ffd5<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6592.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6592.dmg

**Comment by huffman** [comment]
That build error is a bug on VS2012, looking into a fix

**Comment by imgntn** [comment]
it's definitely more robust but a quick two or three refreshes in a row still gets me a crash

```>   Qt5Core.dll!66cd23c9()  Unknown
    [Frames below may be incorrect and/or missing, no symbols loaded for Qt5Core.dll]   
    Qt5Core.dll!66cd3687()  Unknown
    interface.exe!VrMenu::insertAction(QAction * before, QAction * action) Line 169 C++
    [External Code] 
    interface.exe!VrMenu::executeOrQueue(std::function<void __cdecl(VrMenu *)> f) Line 67   C++
    interface.exe!Menu::addActionToQMenuAndActionHash(MenuWrapper * destinationMenu, const QString & actionName, const QKeySequence & shortcut, const QObject * receiver, const char * member, QAction::MenuRole role, int menuItemLocation) Line 659   C++
    interface.exe!Menu::addMenuItem(const MenuItemProperties & properties) Line 963 C++
    interface.exe!Menu::qt_static_metacall(QObject * _o, QMetaObject::Call _c, int _id, void * * _a) Line 117   C++
    Qt5Core.dll!66cfe122()  Unknown
    Qt5Core.dll!66cfd23c()  Unknown
    Qt5Core.dll!66da853b()  Unknown
    Qt5Core.dll!66cdabfb()  Unknown
    Qt5Core.dll!66cfd1a1()  Unknown
    Qt5Core.dll!66cdc4cc()  Unknown
    Qt5Widgets.dll!00af2779()   Unknown
    Qt5Widgets.dll!00af1750()   Unknown
    Qt5Core.dll!66cdabfb()  Unknown
    Qt5Core.dll!66cdc377()  Unknown
    Qt5Core.dll!66d1c29f()  Unknown
    qwindows.dll!5006bd29() Unknown
    Qt5Core.dll!66d1b8f6()  Unknown
    [External Code] 
    Qt5Core.dll!66d1b196()  Unknown
    [External Code] 
    Qt5Gui.dll!63ef4dcf()   Unknown
    [External Code] 
    Qt5Core.dll!66da853b()  Unknown
    [External Code] 
    Qt5Core.dll!66b9abc2()  Unknown
    Qt5Core.dll!66cd78cd()  Unknown
    Qt5Core.dll!66cd7ecf()  Unknown
    Qt5Core.dll!66cd9d10()  Unknown
    Qt5Gui.dll!63d6ff0f()   Unknown
    Qt5Gui.dll!63d6ffe7()   Unknown
    Qt5Gui.dll!63d6de69()   Unknown
    [External Code] 
```

**Comment by huffman** [comment]
My comment was incorrect, I was looking at a different error. The way _stoppingAllScripts was initialized was invoking the copy ctor, which is explicitly deleted on atomic_bool. Hopefully this fixes it (I wasn't seeing the error locally)

**Comment by hifi-gustavo** [comment]
Build console output for commit: faf383ee4dc42cf7a9810abd8b93b7fe6fd80bdf available at: https://gist.github.com/8e0d44e10fddb7f4ffd5<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6592.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6592.dmg

**Comment by hyperlogic** [comment]
I could be wrong, but on x86 bools are atomic anyway

**Comment by huffman** [comment]
@hyperlogic I wasn't so sure at first, but there's a few good Stack Overflow answers to the question: http://stackoverflow.com/questions/16320838/when-do-i-really-need-to-use-atomicbool-instead-of-bool

This is a good one:

> C++'s atomic types deal with three potential problems. First, a read or write can be torn by a task switch if the operation requires more than one bus operation (and that can happen to a bool, depending on how it's implemented). Second, a read or write may affect only the cache associated with the processor that's doing the operation, and other processors may have a different value in their cache. Third, the compiler can rearrange the order of operations if they don't affect the result (the constraints are a bit more complicated, but that's sufficient for now).

**Comment by huffman** [comment]
@imgntn Yeah, I'm still seeing a few crashes on the release build but not on a debug build. I will keep looking for other bugs.

**Comment by hifi-gustavo** [comment]
Build console output for commit: faf383ee4dc42cf7a9810abd8b93b7fe6fd80bdf available at: https://gist.github.com/8e0d44e10fddb7f4ffd5<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6592.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6592.dmg

**Comment by huffman** [comment]
@imgntn Ok, fixed a few other bugs. There's an issue with VrMenu occasionally, but this PR is focused on issues directly related to ScriptEngine.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5f60aeef2217e5fc1e3be0da0754a80982f933a4 available at: https://gist.github.com/8e0d44e10fddb7f4ffd5<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/High+Fidelity-PR-6592-win32.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6592.dmg

File: interface/src/Application.cpp [hunk 1]
```diff
@@ -4317,55 +4324,86 @@ void Application::handleScriptLoadError(const QString& scriptFilename) {
...
+                break;
+            }
+        }
+    }
+    postLambdaEvent([this, scriptName]() {
         _runningScriptsWidget->scriptStopped(scriptName);
         _runningScriptsWidget->setRunningScripts(getRunningScripts());
-    }
+    });
 }
```
Comment (AndrewMeadows) [inline interface/src/Application.cpp]: The **removed** variable appears to be unused.  I wonder if the compiler generates a warning about it.

Comment (huffman) [inline interface/src/Application.cpp]: Thanks for pointing that out, it was a bad merge, it should be used.

**Pull request merged**